File: go/days/day04/part2.go
package day04

func winAmt(winningAmt []int) int {
	var total int

	totalWins := make([]int, len(winningAmt))
	for i := range totalWins {
		totalWins[i] = 1
	}

	for i, amt := range winningAmt {
		for j := 1; j <= amt && i+j < len(winningAmt); j++ {
			totalWins[i+j] += totalWins[i]
		}
	}

	for _, count := range totalWins {
		total += count
	}

	return total
}

File: zig/days/day04/part2.zig
const std = @import("std");

pub fn winAmt(winningAmt: []usize) anyerror!usize {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const alloc = gpa.allocator();
    defer {
        if (gpa.deinit() == .leak) {
            @panic("dealloc failed");
        }
    }

    var totalWins = std.ArrayList(usize).init(alloc);
    defer totalWins.deinit();

    var idx: usize = 0;
    while (idx < winningAmt.len) : (idx += 1) {
        try totalWins.append(1);
    }

    for (winningAmt, 0..) |amt, i| {
        var j: usize = 1;
        while (j <= amt and i + j < winningAmt.len) : (j += 1) {
            totalWins.items[i + j] += totalWins.items[i];
        }
    }

    var total: usize = 0;

    for (totalWins.items) |count| {
        total += count;
    }

    return total;
}
